File: internal/game/states/common_guild.go
package states

import (
	"fmt"

	"github.com/slycrel/slycrel/internal/game"
	"github.com/slycrel/slycrel/internal/mechanics"
	"github.com/slycrel/slycrel/internal/model"
)

// CommonGuildState shows the guild ANSI menu.
// Original: ST_CommonGuild
type CommonGuildState struct{}

func (CommonGuildState) ID() game.StateID { return "common_guild" }

func (CommonGuildState) Enter(s *game.Session) {
	s.Character.Location = model.TheCommonGuild
	s.IO.ShowANSIFile("common_guild_menu")
	s.SetNextState("guild_prompt")
}

// GuildPromptState shows the guild menu.
// Original: ST_GuildShortPrompt + ST_GuildMenu
type GuildPromptState struct{}

func (GuildPromptState) ID() game.StateID { return "guild_prompt" }

func (GuildPromptState) Enter(s *game.Session) {
	s.IO.Outln("[L]evel Up, [Q]uit, [?]Help", true, 1)
	s.IO.Cr()

	choice := s.IO.LettersPrompt("Your choice?", "LQ?", 1, true, true)

	switch choice {
	case "L":
		s.SetNextState("try_new_level")
	case "Q":
		s.Character.Location = model.TheTown
		s.SetNextState("town")
	case "?":
		s.SetNextState("common_guild")
	default:
		s.SetNextState("guild_prompt")
	}
}

// TryNewLevelState attempts to advance the character's level.
// Original: ST_TryNewLevel
type TryNewLevelState struct{}

func (TryNewLevelState) ID() game.StateID { return "try_new_level" }

func (TryNewLevelState) Enter(s *game.Session) {
	c := s.Character
	diff := s.TownConfig.LevelUpDifficulty
	if diff < 1 {
		diff = 4
	}

	// Current level and next level requirement
	currentLevel := c.Level()
	required := mechanics.NextLevelUp(currentLevel+1, diff)

	s.IO.Cr()
	s.IO.Outln(fmt.Sprintf("  Your Current Level: %d", currentLevel), true, 4)
	s.IO.Outln(fmt.Sprintf("  Your Total Experience: %d", c.TotalExperience), true, 4)
	s.IO.Outln(fmt.Sprintf("  Required Experience: %d", required), true, 4)
	s.IO.Cr()

	if currentLevel >= 60 {
		s.IO.Outln("The GuildMaster uncomfortably says, \"I'm sorry, you've maxed out.", true, 2)
	} else if c.TotalExperience >= required {
		s.IO.Outln("\"Finally someone who is ready to go up,\" grins the GuildMaster.", true, 5)
		mechanics.GiveNewLevel(c)
		s.IO.Outln(fmt.Sprintf("You have gained one level! You are now level %d!", c.Level()), true, 3)
		s.IO.Cr()

		// Show new stats
		s.IO.Outln(fmt.Sprintf("  HP: %d  Str: %d  Dex: %d  Spd: %d",
			c.MaxHP, c.Strength, c.Dexterity, c.Speed), true, 1)
		if c.CharClass == model.ClassMage {
			s.IO.Outln(fmt.Sprintf("  Psyche: %d", c.MaxPsyche), true, 1)
		}
		s.Store.SaveCharacter(c)
	} else {
		s.IO.Outln("The GuildMaster looks at you and Laughs.", true, 6)
		needed := required - c.TotalExperience
		s.IO.Outln(fmt.Sprintf("  (You need %d more experience)", needed), true, 1)
	}

	s.IO.Cr()
	s.SetNextState("guild_prompt")
}

File: clients/web/game/states/common_guild.js
import { WhereType, CharClass } from '../../model/enums.js';
import { level } from '../../model/character.js';
import { nextLevelUp, giveNewLevel } from '../../mechanics/leveling.js';
import { saveCharacter } from '../../store/remote.js';
import { TownState } from './town.js';

// The Go server reads this from TownConfig at runtime; we don't have a
// server-side config, so hardcode the same fallback default (4).
const LEVEL_UP_DIFFICULTY = 4;
const MAX_LEVEL = 60;

export class CommonGuildState {
  async enter(session) {
    session.character.location = WhereType.TheCommonGuild;
    session.io.clear();
    await session.io.showAnsiFile('common_guild_menu');
    session.setNext(new GuildPromptState());
  }
}

export class GuildPromptState {
  async enter(session) {
    const { io } = session;
    io.println('[L]evel Up, [Q]uit, [?]Help', 1);
    io.cr();
    const choice = await io.lettersPrompt('Your choice?', 'LQ?');
    io.cr();
    switch (choice) {
      case 'L': session.setNext(new TryNewLevelState()); return;
      case 'Q':
        session.character.location = WhereType.TheTown;
        session.setNext(new TownState());
        return;
      case '?': session.setNext(new CommonGuildState()); return;
    }
    session.setNext(new GuildPromptState());
  }
}

export class TryNewLevelState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    const currentLevel = level(c);
    const required = nextLevelUp(currentLevel + 1, LEVEL_UP_DIFFICULTY);

    io.cr();
    io.println(`  Your Current Level: ${currentLevel}`, 4);
    io.println(`  Your Total Experience: ${c.totalExperience}`, 4);
    io.println(`  Required Experience: ${required}`, 4);
    io.cr();

    if (currentLevel >= MAX_LEVEL) {
      io.println("The GuildMaster uncomfortably says, \"I'm sorry, you've maxed out.", 2);
    } else if (c.totalExperience >= required) {
      io.println('"Finally someone who is ready to go up," grins the GuildMaster.', 5);
      giveNewLevel(c);
      io.println(`You have gained one level! You are now level ${level(c)}!`, 3);
      io.cr();
      io.println(`  HP: ${c.maxHP}  Str: ${c.strength}  Dex: ${c.dexterity}  Spd: ${c.speed}`, 1);
      if (c.charClass === CharClass.Mage) {
        io.println(`  Psyche: ${c.maxPsyche}`, 1);
      }
      await saveCharacter(c);
    } else {
      io.println('The GuildMaster looks at you and Laughs.', 6);
      const needed = required - c.totalExperience;
      io.println(`  (You need ${needed} more experience)`, 1);
    }
    io.cr();
    session.setNext(new GuildPromptState());
  }
}
